judge/forms.py

```diff
@@ -133,16 +136,70 @@ class TOTPForm(Form):
 
     totp_token = NoAutoCompleteCharField(validators=[
         RegexValidator('^[0-9]{6}$', _('Two Factor Authentication tokens must be 6 decimal digits.')),
-    ])
+    ], required=False)
+    webauthn_response = forms.CharField(widget=forms.HiddenInput(), required=False)
 
     def __init__(self, *args, **kwargs):
-        self.totp_key = kwargs.pop('totp_key')
-        super(TOTPForm, self).__init__(*args, **kwargs)
+        self.profile = kwargs.pop('profile')
+        super().__init__(*args, **kwargs)
 
-    def clean_totp_token(self):
-        if not pyotp.TOTP(self.totp_key).verify(self.cleaned_data['totp_token'], valid_window=self.TOLERANCE):
+    def clean(self):
+        if (not self.cleaned_data.get('totp_token') or
+                not pyotp.TOTP(self.profile.totp_key).verify(self.cleaned_data['totp_token'],
+                                                             valid_window=self.TOLERANCE)):
+            raise ValidationError(_('Invalid Two Factor Authentication token.'))
+
+
+class TwoFactorLoginForm(TOTPForm):
+    webauthn_response = forms.CharField(widget=forms.HiddenInput(), required=False)
+
+    def __init__(self, *args, **kwargs):
+        self.webauthn_challenge = kwargs.pop('webauthn_challenge')
+        self.webauthn_origin = kwargs.pop('webauthn_origin')
+        super().__init__(*args, **kwargs)
+
+    def clean(self):
+        if self.profile.is_webauthn_enabled and self.cleaned_data.get('webauthn_response'):
+            if len(self.cleaned_data['webauthn_response']) > 65536:
+                raise ValidationError(_('Invalid WebAuthn response.'))
+
+            if not self.webauthn_challenge:
+                raise ValidationError(_('No WebAuthn challenge issued.'))
+
+            response = json.loads(self.cleaned_data['webauthn_response'])
+            try:
+                credential = self.profile.webauthn_credentials.get(cred_id=response.get('id', ''))
+            except WebAuthnCredential.DoesNotExist:
+                raise ValidationError(_('Invalid WebAuthn credential ID.'))
+
+            user = credential.webauthn_user
+            # Work around a useless check in the webauthn package.
+            user.credential_id = credential.cred_id
+            assertion = webauthn.WebAuthnAssertionResponse(
+                webauthn_user=user,
+                assertion_response=response.get('response'),
+                challenge=self.webauthn_challenge,
+                origin=self.webauthn_origin,
+                uv_required=False,
+            )
+
+            try:
+                sign_count = assertion.verify()
+            except Exception as e:
+                raise ValidationError(str(e))
+
+            credential.counter = sign_count
+            credential.save(update_fields=['counter'])
+
+            return
```

Comment:
else...?

Reply:
Why do you need an else? It just falls through to the TOTP part.